aten/src/ATen/native/LossNLL.cpp

```diff
@@ -475,34 +484,92 @@ Tensor cross_entropy_loss_prob_target(
 
     switch (reduction) {
       case Reduction::Mean:
-        return -(input * target * weight_).sum() / (input.numel() / input.size(1));
+        return -(input * weight_).sum() / (input.numel() / input.size(1));
       case Reduction::Sum:
-        return -(input * target * weight_).sum();
+        return -(input * weight_).sum();
       case Reduction::None:
-        return -(input * target * weight_).sum(1);
+        return -(input * weight_).sum(1);
       default:
         TORCH_CHECK(false, "Invalid reduction type encountered in cross_entropy: ", reduction);
     }
   } else {
     switch (reduction) {
       case Reduction::Mean:
-        return -(input * target).sum() / (input.numel() / input.size(1));
+        return -input.sum() / (input.numel() / input.size(1));
       case Reduction::Sum:
-        return -(input * target).sum();
+        return -input.sum();
       case Reduction::None:
-        return -(input * target).sum(1);
+        return -input.sum(1);
       default:
         TORCH_CHECK(false, "Invalid reduction type encountered in cross_entropy: ", reduction);
     }
   }
 }
 
+Tensor cross_entropy_loss_label_smoothing(
+    const Tensor& self,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction,
+    int64_t ignore_index,
+    double label_smoothing) {
+
+    auto input = at::log_softmax(self, 1, self.scalar_type());
+    auto nllloss = at::nll_loss_nd(input, target, weight, reduction, ignore_index);
+    auto n_classes = input.size(1);
+    auto batches = input.size(0);
+
+    Tensor smooth_loss;
+
+    if (weight.defined()) {
+      // Expand weight to the correct number of dims for broadcasting with input / target
+      auto weight_broadcast_shape = SmallBuffer<int64_t, 5>(input.dim());
+      std::fill(weight_broadcast_shape.begin(), weight_broadcast_shape.end(), 1);
+      weight_broadcast_shape[1] = weight.size(0);
+      Tensor weight_ = weight.view(weight_broadcast_shape);
+
+      input *= weight_;
+    }
+    smooth_loss = -at::sum(input, 1);
+
+    if (ignore_index >= 0) {
+      auto ignore_mask = target.eq(ignore_index);
+      smooth_loss.masked_fill_(ignore_mask, 0.0);
+    }
```

Comment:
```suggestion
    smooth_loss.index_put_({target == ignore_index}, 0.0);
```

Reply:
I think we still need the `if (ignore_index >= 0)` to avoid the operation for the the default, `ignore_index=-100`, case?

Reply:
Hm, I think the kernels for `NLLLoss` just do an unconditional comparison, but I might be wrong.